common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java

```diff
@@ -136,7 +136,11 @@
   private DB db;
 
   public YarnShuffleService() {
-    super("spark_shuffle");
+    this("spark_shuffle");
+  }
+
+  protected YarnShuffleService(String serviceName) {
```

Comment:
So the name by itself isn't going to be enough.  If you really want it configurable we are going to have to have the port configurable. For instance the config name for the port spark.shuffle.service.port needs to be able to be something like spark.shuffle.service.{serviceName}.port.  Otherwise all the spark shuffle servers will try to get the same port and fail.  The only other option will be to use 0 for ephemeral but

Reply:
The name specified here is actually useful only in tests. YARN's service instantiation logic wouldn't even pass the name of the service used in the config to instantiated service. I guess that's the main reason the names and ports are hardcoded or bound to non-namespaced configuration keys.
The way HDP overcomes that is by providing different classpaths with different implementations for different versions of the service (`spark_shuffle` for Spark 1.6.x and `spark2_shuffle` for Spark 2+). The only way I see it's possible to pass different parameters to the same implementation of the service is by providing different configs on the classpath.

I will add a comment here stating that the name is actually only used for the tests, but otherwise would always be hardcoded to `spark_shuffle`.

Reply:
I think there are a few things getting muddled together here -- one is how you'd support running two shuffle services, and the other is how a client could choose which shuffle service it talks to.

The client can already set the port for the shuffle server with `spark.shuffle.service.port`, it just can't set the name used in the `ExecutorRunnable`.

The other thing to add about how the names of the shuffle servers matter in yarn is that the name goes into `yarn-site.xml` as described in the "Configuring the External Shuffle Service" in [`running-on-yarn.md`](https://github.com/apache/spark/blame/c1a5f94973213b1cad15388f3ef8a488424c34a7/docs/running-on-yarn.md#L660).